src/pudl/metadata/classes.py

```diff
@@ -1778,15 +1717,23 @@ def _check_foreign_keys(cls, value):  # noqa: N805
             raise ValueError(
                 format_errors(*errors, title="Foreign keys", pydantic=True)
             )
-        return value
+        return resources
+
+    @model_validator(mode="after")
+    def _populate_from_resources(self: Self):
+        """Populate Package attributes from similar deduplicated Resource attributes.
 
-    @pydantic.root_validator(skip_on_failure=True)
-    def _populate_from_resources(cls, values):  # noqa: N805
+        Resources and Packages share some descriptive attributes. When building a
+        Package out of a collection of Resources, we want the Package to reflect the
+        union of all the analogous values found in the Resources, but we don't want
+        any duplicates. We may also get values directly from the Package inputs.
+        """
         for key in ("keywords", "contributors", "sources", "licenses"):
-            values[key] = _unique(
-                values[key], *[getattr(r, key) for r in values["resources"]]
-            )
-        return values
+            package_value = getattr(self, key)
+            resource_values = [getattr(resource, key) for resource in self.resources]
+            deduped_values = _unique(package_value, *resource_values)
+            setattr(self, key, deduped_values)
+        return self
```

Comment:
However, this creates an infinite recursion when `validate_assignment` is enabled for the model because of [changes in how model validation works](https://docs.pydantic.dev/latest/migration/#changes-to-validators).

For the moment I've disabled `validate_assignment` but that doesn't seem ideal.